### PR TITLE
yasr: init at 0.6.9

### PR DESCRIPTION
**Pull request by jhhuh**
###### Motivation for this change
Adding a new package (yasr-0.6.9). It is to make NixOS more accessible for visually impaired people.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/audio/yasr/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    ./30_conf
+    ./40_dectalk_extended_chars
+  ]; # taken from the debian yasr package
+  
+  meta = {
+    homepage = "http://yasr.sourceforge.net";
+    description = "A general-purpose console screen reader";
+    longDescription = "Yasr is a general-purpose console screen reader for GNU/Linux and other Unix-like operating systems.";
+    platforms = stdenv.lib.platforms.unix;
+    license = stdenv.lib.licenses.gpl2;
```
Comment (Mic92) [inline pkgs/applications/audio/yasr/default.nix]: Do you want to maintain this package?

File: pkgs/applications/audio/yasr/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
+{stdenv,fetchurl}:
+
+stdenv.mkDerivation rec {
+  name = "yasr-${version}";
+  
+  version = "0.6.9";
+  
+  src = fetchurl {
+    url = "https://sourceforge.net/projects/yasr/files/yasr/0.6.9/yasr-0.6.9.tar.gz";
```
Comment (Mic92) [inline pkgs/applications/audio/yasr/default.nix]: Can be rewritten as:

```
url = "https://sourceforge.net/projects/yasr/files/yasr/${version}/yasr-${version}.tar.gz";
```

File: pkgs/applications/audio/yasr/30_conf [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+--- yasr.orig/yasr.conf	2008-01-31 14:01:50.000000000 +0100
++++ yasr/yasr.conf	2012-01-28 21:43:09.027710410 +0100
+@@ -55,13 +55,13 @@ cursor tracking=full
+ #synthesizer=doubletalk
+ #synthesizer=bns
+ #synthesizer=apollo
+-synthesizer=emacspeak server
+-#synthesizer=speech dispatcher
++#synthesizer=emacspeak server
++synthesizer=speech dispatcher
```
Comment (Mic92) [inline pkgs/applications/audio/yasr/30_conf]: does these configuration changes also make sense for Nixos? Is `speech` to be preferred over `emacspeak`?
 At least the `shell=/bin/bash` should be `shell=/bin/sh` (sh is symlink to bash)

File: pkgs/applications/audio/yasr/default.nix (hunk 1 above)
Comment (jhhuh) [inline pkgs/applications/audio/yasr/default.nix]: Thanks for your review. Yes, I am interested in being maintainer.

File: pkgs/applications/audio/yasr/default.nix (hunk 2 above)
Comment (jhhuh) [inline pkgs/applications/audio/yasr/default.nix]: Yes, you're right.

File: pkgs/applications/audio/yasr/30_conf (hunk 3 above)
Comment (jhhuh) [inline pkgs/applications/audio/yasr/30_conf]: Yes, you're right about 'shell'. To be honest I just took the patch from the debian yasr package. I need to figure out what synthesizer would be preferred in nixos (at least, eflite and freetts do not seem to work out of box).

File: pkgs/applications/audio/yasr/default.nix (hunk 1 above)
Comment (jhhuh) [inline pkgs/applications/audio/yasr/default.nix]: @Mic92 I just put myself in stdenv.lib.maintainers. Is it something allowed?

Comment (Mic92) [inline pkgs/applications/audio/yasr/default.nix]: @jhhuh yes, this is the way to become a maintainer.

File: pkgs/applications/audio/yasr/40_dectalk_extended_chars [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+===================================================================
+--- yasr-0.6.9.orig/yasr/tts.c	2008-02-03 00:10:07.000000000 +1100
++++ yasr-0.6.9/yasr/tts.c	2012-02-05 10:59:06.059007839 +1100
+@@ -281,6 +281,9 @@
+   char *p = synth[tts.synth].unspeakable;
+ 
+   if (ch < 32) return 1;
++  /* characters with high bit set cause DECTALK to crash */
++  if (tts.synth == TTS_DECTALK && ch & 0x80)
++    return 1;
```
Comment (Mic92) [inline pkgs/applications/audio/yasr/40_dectalk_extended_chars]: This patch sounds useful from the description. Do you know if it was reported upstream?

File: pkgs/applications/audio/yasr/10_fix_openpty_forkpty_declarations [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ /* cfmakeraw.c prototypes */
+ extern void cfmakeraw(struct termios *);
+@@ -317,8 +319,10 @@ extern void cfmakeraw(struct termios *);
+ /* login_tty.c prototypes */
+ extern int login_tty(int);
+ 
++#ifndef HAVE_FORKPTY
+ /* forkpty.c prototypes */
+ extern int forkpty(int *, char *, struct termios *, struct winsize *);
++#endif
```
Comment (Mic92) [inline pkgs/applications/audio/yasr/10_fix_openpty_forkpty_declarations]: As well as this one

**Comment by Mic92** [comment]
I do not know about the maintenance state of `yars`, if it is unresponsive then it is also ok to maintain the patches our self.

File: pkgs/applications/audio/yasr/40_dectalk_extended_chars (hunk 4 above)
Comment (jhhuh) [inline pkgs/applications/audio/yasr/40_dectalk_extended_chars]: As I made in a comment, this piece is not written by me. I blindly took it from the YASR debian source package. I just checked that this fix is implemented in the current HEAD version (2013) of YASR [https://sourceforge.net/p/yasr/git/ci/master/tree/yasr/tts.c#l286)]. So I guess the author is well aware of all these patches, and it is just that the official release is stuck at the old version v0.6.9 in 2008.

File: pkgs/applications/audio/yasr/20_maxpathlen [hunk 6]
```diff
@@ -0,0 +1,11 @@
+--- a/yasr/config.c.orig	2011-11-28 03:56:58.764995828 +0100
++++ a/yasr/config.c	2011-11-28 03:57:00.048967703 +0100
+@@ -60,7 +60,7 @@
+   int args, arg[16], *argp;
+   int i, key, ln = 0, mode = 0;
+   char *home, *ptr, *s;
+-  char confname[MAXPATHLEN];
++  char confname[strlen(PACKAGE_DATA_DIR) + 10 + 1];
```
Comment (jhhuh) [inline pkgs/applications/audio/yasr/20_maxpathlen]: @mic92 Same as the above.

File: pkgs/applications/audio/yasr/10_fix_openpty_forkpty_declarations (hunk 5 above)
Comment (jhhuh) [inline pkgs/applications/audio/yasr/10_fix_openpty_forkpty_declarations]: Same as the above :)

**Comment by jhhuh** [comment]
@Mic92 I thought the maintenance of the nix expression (in nixpkgs) for a program is independent from the maintenance of the program itself, is it? That is why I didn't bother to check it at first. In any case, now I asked the author if he is fine with it. We will see.

**Comment by jhhuh** [comment]
@Mic92 The author is still unresponsive to my message. Should we go as is now?

**Comment by Mic92** [comment]
Yes. I just squashed the commits into one.

**Pull request closed without merging**